Remove parsing perf  bottleneck in WordEmbeddingsTransform

adamsitnik opened the pull request:
This PR improves the performance of reading large text files and affects two of our most time-consuming benchmarks.

Info:

```ini
BenchmarkDotNet=v0.11.2, OS=Windows 10.0.17134.345 (1803/April2018Update/Redstone4)
Intel Xeon CPU E5-1650 v4 3.60GHz, 1 CPU, 12 logical and 6 physical cores
Frequency=3507503 Hz, Resolution=285.1031 ns, Timer=TSC
.NET Core SDK=3.0.100-alpha1-009697
  [Host]     : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
  Job-OXDQNP : .NET Core 2.1.5 (CoreCLR 4.6.26919.02, CoreFX 4.6.26919.02), 64bit RyuJIT
```

Before:

|                                         Method |    Mean |
|----------------------------------------------- |--------:|
| WikiDetox_WordEmbeddings_OVAAveragedPerceptron | 286.7 s |
|                WikiDetox_WordEmbeddings_SDCAMC | 184.1 s |

After:

|                                         Method |     Mean |
|----------------------------------------------- |---------:|
| WikiDetox_WordEmbeddings_OVAAveragedPerceptron | 169.02 s |
|                WikiDetox_WordEmbeddings_SDCAMC |  65.32 s |

Which is two minutes less to read the huge file for both benchmarks which results in a **x3** boost for `WikiDetox_WordEmbeddings_SDCAMC`  and **40%** improvement for `WikiDetox_WordEmbeddings_OVAAveragedPerceptron`

Reading the file was a bottleneck:

![image](https://user-images.githubusercontent.com/6011991/48314555-83f82680-e5cb-11e8-8948-cc5a22a07c95.png)

I have applied all possible optimizations and parallelized this operation.

I am going to post a detailed description on Monday.

adamsitnik commented:
/cc @danmosemsft

justinormont commented on test/Microsoft.ML.Benchmarks/Harness/Configs.cs:
```diff
@@ -15,22 +15,21 @@ public RecommendedConfig()
             Add(GetJobDefinition() // job defines how many times given benchmark should be executed
                 .WithCustomBuildConfiguration(GetBuildConfigurationName())
-                .With(CreateToolchain())); // toolchain is responsible for generating, building and running dedicated executable per benchmark
+                .With(CreateToolchain())); // toolchain is responsible for generating, building and running dedicated executable per benchmark);
```
Extraneous ');'

```suggestion
                .With(CreateToolchain())); // toolchain is responsible for generating, building and running dedicated executable per benchmark
```

adamsitnik replied:
The number of `)` is fine, one for `CreateToolchain`, 2nd for `With` and 3rd for `Add` from line 16.

To prevent from such errors and breaking the benchmarks I added an integration test some time ago, so now if the build is green it means that we can run the benchmarks. https://github.com/dotnet/machinelearning/blob/19d907766a809bf0f91da15b5f10056dc2090942/test/Microsoft.ML.Benchmarks.Tests/BenchmarksTest.cs#L47

eerhardt replied:
I believe @justinormont was referring to the end of the comment like.  Comments in C# don't need to end in a semi-colon.  😝

adamsitnik replied:
@justinormont I apology for the misunderstanding! 

@eerhardt  Thanks for pointing this out. It was hilarious ;p

TomFinley commented on src/Microsoft.ML.Transforms/Text/LineParser.cs:
```diff
@@ -0,0 +1,70 @@
+
+namespace Microsoft.ML.Transforms.Text
+{
+    public static class LineParser
```
>LineParser [](start = 24, length = 10)

Ought we to make this internal?

adamsitnik replied:
sure, done!

TomFinley commented on src/Microsoft.ML.Transforms/Text/LineParser.cs:
```diff
@@ -0,0 +1,70 @@
+    {
+        private static readonly char[] _delimiters = { ' ', '\t' };
+
+#if NETSTANDARD2_0
```
>NETSTANDARD2_0 [](start = 4, length = 14)

Such broad divergences in implementation make me a little nervous, since unless I'm very much mistaken we don't actually have multiple configurations of whether we have one version of .NET standard or another. Is this separation absolutely urgent? If it has to do with the presence or absence of methods of span of char, could ReadOnlyMemoryUtilities be made to serve?

danmosemsft commented on src/Microsoft.ML.Transforms/Text/LineParser.cs:
```diff
@@ -0,0 +1,70 @@
+    {
+        private static readonly char[] _delimiters = { ' ', '\t' };
+
+#if NETSTANDARD2_0
```
It seems to me it is worth beginning to multitarget an assembly if there is a significant win (like in this case). However I suggest partial classes rather than #if when dealing with entire methods, like here. That pattern has helped lower code complexity in corefx.

cc @eerhardt for thoughts

TomFinley replied:
That's fine, but I'd first like to understand *why* we think the divergence is necessary, since the divergence seems to increase our required test load. If there's a good reason, that's fine, but my suspicion is that it is mostly due to being unaware of the conveniences we ourselves had to build up to parse text files quickly. (E.g., `DoubleParser.TryParse`), which might all around lead to better performing code in both cases, not one.

adamsitnik replied:
@danmosemsft @TomFinley I had no idea about the existence of `DoubleParser`. I have re-used it and now we have a simple implementation again.

Ivanidzo4ka commented on src/Microsoft.ML.Transforms/Text/WordEmbeddingsTransform.cs:
```diff
@@ -723,60 +725,53 @@ private Model GetVocabularyDictionary()
+                    var parsedData = new ConcurrentBag<(string key, float[] values, long lineNumber)>();
+                    int skippedLinesCount = Math.Max(1, _linesToSkip);
+
+                    Parallel.ForEach(File.ReadLines(_modelFileNameWithPath).Skip(skippedLinesCount),
```
>Parallel.ForEach [](start = 20, length = 16)

Please respect Host.ConcurrencyFactor and set MaxDegreeOfParallelism to it's value.

adamsitnik replied:
@Ivanidzo4ka great catch! thanks! #done

eerhardt commented on src/Microsoft.ML.Transforms/Microsoft.ML.Transforms.csproj:
```diff
@@ -1,7 +1,7 @@
 
   <PropertyGroup>
-    <TargetFramework>netstandard2.0</TargetFramework>
+    <TargetFrameworks>netstandard2.0;netcoreapp2.1</TargetFrameworks>
```
You are going to need to do a lot more work here in order to get our nuget package to be multi-targeted correctly.

NuGet only picks assets from 1 TFM. And since we put many assemblies in `Microsoft.ML`, you are only making this assembly multi-targeted. Which means the NuGet package will have all assemblies in `netstandard2.0` and only this assembly in `netcoreapp2.1`. So if a consumer is targeting netcoreapp2.1, they won't get references to all the other assemblies.

We already have a double parser code in ML.NET that takes in a `ROS<char>`, maybe you can change your code to use that so you don't have to take a dependency on `netcoreapp2.1`-specific APIs?

adamsitnik replied:
@eerhardt  I had no idea about the existence of DoubleParser. I have re-used it and now we target just .NET Standard 2.0 again.

eerhardt commented on test/Microsoft.ML.Benchmarks/Microsoft.ML.Benchmarks.csproj:
```diff
@@ -4,6 +4,7 @@
     <OutputType>Exe</OutputType>
     <LangVersion>7.2</LangVersion>
     <StartupObject>Microsoft.ML.Benchmarks.Program</StartupObject>
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
```
Let's not do this.  What is it required for? We typically don't do this in .NET Core projects (thus why it is not the default)

---
In reply to: [232782005](https://github.com/dotnet/machinelearning/pull/1599#discussion_r232782005) [](ancestors = 232782005)

adamsitnik replied:
@eerhardt  When we do` dotnet run` the MSBuild copies to the output folder the dependencies that are somehow consumed in the code. If we don't use any type from BenchmarkDotNet.Diagnostics.Windows the .dll does not get copied and the console line argument parsing logic for `--profiler ETW` fails to load the .dll in a dynamic way ;(

eerhardt replied:
How does the dynamic loading of `--profiler ETW` work? This sounds like a bug in that logic. You should just be able to load an assembly, and it will load correctly if it is in the `.deps.json` file (which this one will be since you took a PackageReference on it).

eerhardt replied:
BTW - 

> When we do dotnet run the MSBuild copies to the output folder the dependencies that are somehow consumed in the code.

That is not true. Assemblies that are referenced through NuGet packages are not copied to the output folder, whether you use them in code or not. NuGet assemblies are loaded from the NuGet cache during `dotnet run`.

adamsitnik replied:
@eerhardt I created a separate issue in BDN for that: https://github.com/dotnet/BenchmarkDotNet/issues/955

eerhardt commented on test/Microsoft.ML.Benchmarks/Microsoft.ML.Benchmarks.csproj:
```diff
@@ -12,6 +13,8 @@
   <ItemGroup>
     <PackageReference Include="BenchmarkDotNet" Version="$(BenchmarkDotNetVersion)" />
+    <PackageReference Include="BenchmarkDotNet.Diagnostics.Windows" Version="$(BenchmarkDotNetVersion)" />
+    <PackageReference Include="Microsoft.Diagnostics.Tracing.TraceEvent" Version="$(TraceEventVersion)" />
```
`TraceEventVersion` => `MicrosoftDiagnosticsTracingTraceEventVersion`

We name the MSBuild property the same name as the NuGet package it corresponds to.

eerhardt commented on src/Microsoft.ML.Transforms/Text/LineParser.cs:
```diff
@@ -0,0 +1,70 @@
+                if (valuesToParse[i] == ' ' || valuesToParse[i] == '\t')
+                    valuesCount++;
+
+            float[] values = new float[valuesCount + 1]; // + 1 because the line is trimmed and there is no whitespace at the end
```
Alternatively, start `valuesCount` at `1`.

adamsitnik replied:
I decided to put it into a dedicated method. The code is still ugly, but a little bit less now.

shauheen commented:
@adamsitnik can you please file a separate issue for this change and associate the PR. Thanks.

adamsitnik commented:
@shauheen done #1608

danmosemsft commented:
Seems all feedback was addressed? Just needs rebase?

danmosemsft commented:
@shauheen do you have further feedback?

eerhardt commented on test/Microsoft.ML.Benchmarks/Microsoft.ML.Benchmarks.csproj:
```diff
@@ -4,6 +4,7 @@
     <OutputType>Exe</OutputType>
     <LangVersion>7.2</LangVersion>
     <StartupObject>Microsoft.ML.Benchmarks.Program</StartupObject>
+    <CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
```
Now that [https://github.com/dotnet/BenchmarkDotNet/issues/955](https://github.com/dotnet/BenchmarkDotNet/issues/955) has been fixed, can this be removed? I would like to not clutter up the .csproj unnecessarily.

adamsitnik replied:
I am going to remove it as soon as we ship the fix to NuGet (today or tomorrow)

adamsitnik replied:
solved

eerhardt commented:
> using System;

need a copyright header

---
Refers to: src/Microsoft.ML.Core/Utilities/LineParser.cs:1 in 0190b0c. [](commit_id = 0190b0cf6055887ad52f18c4fe18239156885435, deletion_comment = False)

eerhardt commented on src/Microsoft.ML.Core/Utilities/LineParser.cs:
```diff
@@ -0,0 +1,55 @@
+    [BestFriend]
+    internal static class LineParser
+    {
+        [BestFriend] // required because the return type is a value tuple..
```
I'm not sure I understand this comment. Shouldn't the BestFriend attribute on the class be enough?

Maybe keep the class `internal`, but make the method `public`?

/cc @TomFinley

adamsitnik replied:
It seems that it treats a ValueTuple as defined in this library and adding the `[BestFriend]` for the method makes it work. If I remove it, it does not work.

eerhardt replied:
Wow, that's super odd and unexpected IMO.

Even if you make the method `public` it doesn't work?

adamsitnik replied:
hah, with public it works ;)

eerhardt commented on test/Microsoft.ML.Benchmarks/Harness/Configs.cs:
```diff
@@ -18,19 +18,18 @@ public RecommendedConfig()
         /// we need our own toolchain because MSBuild by default does not copy recursive native dependencies to the output
         /// </summary>
-        private IToolchain CreateToolchain()
+        protected IToolchain CreateToolchain()
```
Why does this and `GetBuildConfigurationName` need to be `protected`?

adamsitnik replied:
good catch, solved

eerhardt commented on src/Microsoft.ML.Core/Utilities/LineParser.cs:
```diff
@@ -0,0 +1,55 @@
+                }
+            }
+
+            return (true, trimmedLine.Slice(0, firstSeparatorIndex).ToString(), values);
```
I'm not sure how important this is (but since we are doing this in the name of perf):

The `string key` allocates a new string here, and later is used to add to a `NormStr.Pool`, which can operate over `ReadOnlyMemory<char>`.

If we returned a `ROM<char>` here instead, we wouldn't need to allocate a new string at all.

adamsitnik replied:
I did not show in the profiles. I will give it a try in round 2 of the optimizations.

eerhardt commented on src/Microsoft.ML.Transforms/Text/WordEmbeddingsTransform.cs:
```diff
@@ -723,62 +725,64 @@ private Model GetVocabularyDictionary()
+
+                    // Handle first line of the embedding file separately since some embedding files including fastText have a single-line header
+                    var firstLine = File.ReadLines(_modelFileNameWithPath).First();
+                    string[] wordsInFirstLine = firstLine.TrimEnd().Split(' ', '\t');
```
Why doesn't this call `LineParser.ParseKeyThenNumbers`?

adamsitnik replied:
because `ParseKeyThenNumbers` expects that every line is sth like `string float float float float ...`. If the first line of some file would be `string string string` the current code would return dimension = 2, the `ParseKeyThenNumbers` would return 0. It's a single line so there would be no noticeable perf gain and I did not want to risk.

eerhardt commented on test/Microsoft.ML.Tests/Transformers/LineParserTests.cs:
```diff
@@ -0,0 +1,40 @@
+        {
+            foreach (var line in new string[]
+            {
+                "key 0.1 0.2 0.3", "key 0.1 0.2 0.3 ",
```
I would make this 4 lines instead of 2. The way it reads is that there are 2 lines, but you have to parse the lines to see that there are really 4.

eerhardt commented:
> using Microsoft.ML.Runtime.Internal.Utilities;

Add copyright header.

---
Refers to: test/Microsoft.ML.Tests/Transformers/LineParserTests.cs:1 in 0190b0c. [](commit_id = 0190b0cf6055887ad52f18c4fe18239156885435, deletion_comment = False)

adamsitnik commented:
@shauheen @eerhardt I have addressed all issues, PTAL one more time.

@tannergooding thanks for pointing the `Utf8Parser`. I did not use it in this PR because I don't know if every input file is `Utf8`. In the next PR I will provide "fast path" for UTF8 files.

eerhardt commented:
> using System;

Still needs a copyright header.

---
In reply to: [439979964](https://github.com/dotnet/machinelearning/pull/1599#issuecomment-439979964) [](ancestors = 439979964)

---
Refers to: src/Microsoft.ML.Core/Utilities/LineParser.cs:1 in 0190b0c. [](commit_id = 0190b0cf6055887ad52f18c4fe18239156885435, deletion_comment = False)
